File: OpenX/dotnet/Engines/OpenX.Engine.EginX/Engine_EginX.cs
using EginX;
using OpenX.Client;
using OpenX.Gfx;
using System;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
#pragma warning disable CS0649, CS0169

[assembly: InternalsVisibleTo("OpenX.GfxTests")]

namespace OpenX;

#region Client

public class ExClientHost : Game, IClientHost {
    public ClientBase Client;
    public SceneBase Scene;
    public IPluginHost PluginHost;

    public ExClientHost(Func<ClientBase> client) {
        Client = client();
        DeviceManager = new GraphicsDeviceManager(this);
        //PluginHost = pluginHost; IPluginHost pluginHost, string title
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public T GetScene<T>() where T : SceneBase => Scene as T;

    public void SetScene(SceneBase scene) { Scene?.Dispose(); Scene = scene; Scene?.Load(); }

    protected override async Task LoadContent() {
        await base.LoadContent();
        await Client.LoadContent();
    }

    protected override async Task UnloadContent() {
        await Client.UnloadContent();
        await base.UnloadContent();
    }
}


#endregion

#region Engine

// EginXObjectBuilder
// MISSING

// EginXGfxSprite2D
public class EginXGfxSprite2D : IOpenGfxSprite<object, object> {
    readonly SpriteManager<object> _spriteManager;
    public EginXGfxSprite2D() {
        //_spriteManager = new SpriteManager<Sprite2D>(new GodotSpriteBuilder());
    }

    public SpriteManager<object> SpriteManager => _spriteManager;
    public void Preload(ISource source, object path) => _spriteManager.Preload(source, path);
    public Task<(object spr, object tag)> Create(ISource source, object path, object parent = null) => _spriteManager.Create(source, path);
}

// EginXEngine
public class EginXEngine : Engine {
    public static readonly Engine This = new EginXEngine();
    EginXEngine() : base("EX", "EginX") {
        Caps = EngineX.Caps.Drawing;
        GfxFactory = () => [null, new EginXGfxSprite2D(), null, null, null, null];
        SfxFactory = () => [new SystemSfx()];
    }
}

#endregion

File: OpenX/python/openx/engines/eginx/engine_eginx.py
from __future__ import annotations
import traceback
from openx.core import ISource, Engine
from openx.engines.system import SystemSfx
from openx.gfx import IOpenGfxApi, IOpenGfxSprite, IOpenGfxModel, ObjectModelBuilder, ObjectModelManager, MaterialBuilder, MaterialManager, ShaderBuilder, ShaderManager, TextureManager, TextureBuilder
from openx.client import IClientHost
from openx.gfx.egin import Game, GraphicsDeviceManager

#region Client

# EginXClientHost
class EginXClientHost(Game, IClientHost):
    def __init__(self, client: callable):
        super().__init__()
        # self.deviceManager = GraphicsDeviceManager(self)
        self.client: ClientBase = client()
        self.scene: SceneBase = None
        self.pluginHost: IPluginHost = None

    def getScene[T](self) -> T: return self.scene

    def setScene(self, scene: SceneBase) -> None:
        if self.scene: self.scene.dispose()
        self.scene = scene
        self.scene.load() 

    def loadContent(self) -> None:
        super().loadContent()
        self.client.loadContent()

    def unloadContent(self) -> None:
        self.client.unloadContent()
        super().unloadContent()

#endregion

#region Engine

# EginXObjectBuilder
# MISSING

# EginXGfxApi
class EginXGfxApi(IOpenGfxApi):
    def __init__(self): pass
    def attach(self, method: GfxAttach, src: object, args: list[object]) -> object: raise NotImplementedError()

# EginXGfxSprite2D
class EginXGfxSprite2D(IOpenGfxSprite):
    def __init__(self):
        self.spriteManager: SpriteManager = SpriteManager(OpenGLTextureBuilder())
    def preload(self, source: ISource, path: object) -> None: self.spriteManager.preload(path)
    def create(self, source: ISource, path: object, level: range = None) -> int: return self.spriteManager.create(path)[0]

# EginXEngine
class EginXEngine(Engine):
    def __init__(self):
        super().__init__('EX', 'EginX')
        self.gfxFactory = staticmethod(lambda: [EginXGfxApi(), EginXGfxSprite2D(), None, None, None, None])
        self.sfxFactory = staticmethod(lambda: [SystemSfx()])
EginXEngine.this = EginXEngine()

#endregion